draft-ietf-jsonpath-base.md

```diff
@@ -1153,17 +1153,24 @@ resulting in an empty nodelist, or
 When any path on either side of a comparison results in a nodelist consisting of a single node, each such path is
 replaced by the value of its node and then:
 
-* comparison using one of the operators `==` yields true if and only if the comparison
-is between equal primitive values.
+* a comparison using the operator `==` yields true if and only if the comparison
+is between equal primitive values (numbers, strings, `true`, `false`, or `null`).
 
-* comparisons using one of the operators `<=` or `>=` yield true if and only if
-the comparison is between numeric values which satisfy the comparison.
+* a comparison using the operator `!=` yields true if and only if the comparison
+is not between equal primitive values (numbers, strings, `true`, `false`, or `null`).
 
-* any comparison of two values using one of the operators `!=`, `>`, `<` is defined as the negation of the comparison
-of the same values using the operator `==`, `<=`, `>=`, respectively.
+* a comparison using one of the operators `<`, `<=`, `>`, or `>=` yields true if and only if
+the comparison is between numbers, or between strings, which satisfy the comparison:
 
-Note that `==` comparisons between a structured value and any value, including the same structured value, yield false.
-Consequently, `!=` comparisons between a structured value and any value, including the same structured value, yield true.
+    * numbers in the I-JSON {{-i-json}} range of exact values MUST compare using the normal mathematical ordering;
+      one or both numbers outside that range MAY compare using an implementation specific ordering
+    * the empty string compares less than any non-empty string
+    * a non-empty string compares less than another non-empty string if and only if the first string starts with a
```

Comment:
Let's go for scalar values and, perhaps for the benefit of non-experts like me, point out that these do not include surrogates.

Where do we say strings in the JSON value must be in UTF-8?

Yes, "code point" rather than, ahem, "codepoint".

Reply:
Urgh, I meant “code point” (2 words) not “codepoint” (compound word).  See https://www.unicode.org/versions/Unicode10.0.0/ch03.pdf#G7404

Reply:
> Where do we say strings in the JSON value must be in UTF-8?

They aren't -- they are sequences of Unicode Scalar Values.
(They are in UTF-8 in the JSON text.)

Reply:
We should say that?

Reply:
> scalar values do not include UTF-16 surrogates

This brings up an interesting point.  JSON requires that UTF-16 characters be encoded in `\uxxxx` pairs.  Is the encoding sufficient to determine correct ordering (i.e. does sorting by the encoded strings yield the same result as sorting the unencoded characters)?

I'd like to see (maybe in a follow-up clarification PR) some language around this.

Reply:
I've adopted the term Unicode scalar value in fabec5237abf9d35242d7736354d81fd1cf02873 and moved it to the terminology section in 112e8aa63a546db66563f9d3d42dec9d0e33ecdd (to make it clear that "value" in the term "Unicode scalar value" is distinct from our term "value", meaning JSON value.)

(If the spec needs to describe encoding, that should go in a separate PR.)